src/Handlers.jl

```diff
@@ -110,37 +148,78 @@ function splitsegments(r::Router, h::Handler, segments)
     end
     return vals
 end
+function newsplitsegments(segments)
+    vals = Expr[]
+    for s in segments
+        if s == "*" #TODO: or variable, keep track of variable types and store in handler
+            T = Any
+        else
+            v = Val(Symbol(s))
+            T = typeof(v)
+        end
+        push!(vals, Expr(:(::), T))
+    end
+    return vals
+end
+
+function gethandler end
+gethandler(r::Router, args...) = r.default
 
-function register!(r::Router, method::DataType, scheme, host, path, handler)
+function register!(r::Router{id}, method, scheme, host, path, handler) where {id}
+    Base.depwarn("`HTTP.register!(r::Router, ...)` is deprecated, use `HTTP.@register r ...` instead", nothing)
     # save string => Val mappings in r.segments
     segments = map(String, split(path, '/'; keepempty=false))
-    vals = splitsegments(r, handler, segments)
+    vals = splitsegments(r, segments)
     # return a method to get dispatched to
     #TODO: detect whether defining this method will create ambiguity?
-    @eval $(r.sym)(::$method, ::$scheme, ::$host, $(vals...), args...) = $handler
+    @eval gethandler(r::Router{$(Meta.QuoteNode(id))}, ::$method, ::$scheme, ::$host, $(vals...), args...) = $handler
     return
 end
 
-function gethandler(r::Router, req)
+gh(s::String) = isempty(s) ? Any : typeof(Val(Symbol(s)))
+gh(s::Symbol) = s
+
+function generate_gethandler(router, method, scheme, host, path, handler)
+    vals = :(HTTP.Handlers.newsplitsegments(map(String, split($path, '/'; keepempty=false)))...)
+    q = esc(quote
+        $(router).routes[HTTP.Handlers.Route($method, $scheme, $host, $path)] = $(string(handler))
+        @eval function HTTP.Handlers.gethandler(r::$(Expr(:$, :(typeof($router)))),
```

Comment:
Calling `@eval` from within a quote, from with in a macro,
can normally be avoided with some thought and `Meta.quote`, use.